test/functional/p2p_eviction.py

```diff
@@ -0,0 +1,141 @@
+            block_peer.sync_with_ping(timeout=5)
+            best_block = node.getbestblockhash()
+            tip = int(best_block, 16)
+            best_block_time = node.getblock(best_block)['time']
+            block = create_block(tip, create_coinbase(node.getblockcount() + 1), best_block_time + 1)
+            block.solve()
+            block_peer.send_blocks_and_test([block], node, success=True)
+            self.protected_peers.add(current_peer)
+
+        # The next 5 nodes are slow-pinging peers, making them eviction candidates.
+        for _ in range(5):
+            node.add_p2p_connection(SlowP2PInterface())
+            current_peer += 1
+
+
+        #The next 4 peers send us a tx, protecting them from eviction
+        for i in range(4):
+            txpeer = node.add_p2p_connection(SlowP2PInterface())
+            current_peer += 1
+            txpeer.sync_with_ping(timeout=5)
+
+            prevtx = node.getblock(node.getblockhash(i+1), 2)['tx'][0]
+            rawtx = node.createrawtransaction(
+                inputs=[{'txid': prevtx['txid'], 'vout': 0}],
+                outputs=[{node.get_deterministic_priv_key().address: 50 - 0.00125}],
+            )
+            sigtx = node.signrawtransactionwithkey(
+                hexstring=rawtx,
+                privkeys=[node.get_deterministic_priv_key().key],
+                prevtxs=[{
+                    'txid': prevtx['txid'],
+                    'vout': 0,
+                    'scriptPubKey': prevtx['vout'][0]['scriptPubKey']['hex'],
+                }],
+            )['hex']
+            txpeer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
+            self.protected_peers.add(current_peer)
+
+        # The next 8 peers have faster pings, which will usually protect them from eviction.
+        for _ in range(8):
+            fastpeer = node.add_p2p_connection(P2PInterface())
+            current_peer += 1
+            wait_until(lambda: "ping" in fastpeer.last_message, timeout=10)
+
+        # Make sure by asking the node what the actual min pings are
+        peerinfo = node.getpeerinfo()
+        pings = [[0 for i in range(2)] for j in range(len(node.p2ps))]
+
+        # After adding 21 peers, the next one hits the maxconnection limit and triggers the eviction mechanism.
+        node.add_p2p_connection(SlowP2PInterface())
+
+        # Usually the 8 fast peers are protected. In rare case of unreliable pings,
+        # one of the slower peers might have a faster min ping though.
+        for i in range(len(peerinfo)):
+            pings[i][0] = i
+            pings[i][1] = peerinfo[i]['minping'] if 'minping' in peerinfo[i] else 1000000
+
+        sorted_pings = sorted(pings, key = lambda x:x[1])
+
+        for i in range(8):
+            self.protected_peers.add(sorted_pings[i][0])
+
+        # One of the non-protected peers must be evicted. We can't be sure which one because
+        # 4 peers are protected via netgroup, which is identical for all peers,
```

Comment:
This is coming a little confusing for me, which 4 peers are protected via netgroup? If i understand correctly for functional testing all the peers are from same netgroup. So there's no way to differentiate between them via IP addresses. Is this what this comment is trying to convey?

Reply:
I believe these comments above try to convey this behaviour https://github.com/bitcoin/bitcoin/pull/16756/files#diff-96b7763a95e093580bd4507bce07ff22R10-R12

Reply:
I think it is impossible to say which peers are protected via netgroup because it can vary from run to run! 
Since A) this is the first check in `AttemptToEvictConnection()`  B) all test nodes have identical netgroup and C) `std::sort` does not guarantee to preserve the order of identical elements, this step should protect 4 arbitrary peers in the test. 
If, for example, all 4 block-sending peers would be protected by  chance, then the later `EraseLastKElements` with `CompareNodeBlockTime` would operate on identical elements and erase 4 random peers from `vEvictionCandidates` (protecting them). 
The assertion at the end of the tests only tests that no peer assumed to be protected will be evicted. Its protection might happen in the `EraseLastKElements` specific to its unique feature, or earlier by chance.

Reply:
Yes, the only certainly we have regarding the evicted peers in this test is that it will be one of the zero-indexed peers in the set of {4, 5, 6, 7, 8} e.g. the slow-pinging ones.